javatests/arcs/core/data/RecipeTest.kt

```diff
@@ -0,0 +1,284 @@
+import com.google.common.truth.Truth.assertThat
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+/** Tests for [Recipe]. */
+@RunWith(JUnit4::class)
+internal class RecipeTest {
+
+    // Type: [Person {name: Text}]
+    private val personCollectionType = CollectionType(EntityType(Schema(
+        names = setOf(SchemaName("Person")),
+        fields = SchemaFields(
+            singletons = mapOf("name" to FieldType.Text),
+            collections = mapOf()
+        ),
+        hash = "abcd"
+    )))
+
+    // Type: [Contact {id: Text}]
+    private val contactCollectionType = CollectionType(EntityType(Schema(
+        names = setOf(SchemaName("Contact")),
+        fields = SchemaFields(
+            singletons = mapOf("id" to FieldType.Text),
+            collections = mapOf()
+        ),
+        hash = "efgh"
+    )))
+
+    @Before
+    fun before() {
+        DriverAndKeyConfigurator.configureKeyParsers()
+    }
+
+    @Test
+    fun handleToStorageKey_withStorageKey() {
+        val storageKey = "reference-mode://{db://abcd@arcs/Person}{db://abcd@arcs//handle/people}";
+
+        assertThat(
+            Recipe.Handle(
+                name = "people",
+                fate = Fate.MAP,
+                type = personCollectionType,
+                storageKey = storageKey
+            ).toStorageKey()
+        ).isEqualTo(
+            StorageKeyParser.parse(storageKey)
+        )
+    }
+
+    @Test
+    fun handleToStorageKey_withCapabilities() {
+        assertThat(
+            Recipe.Handle(
+                name = "people",
+                fate = Fate.CREATE,
+                type = personCollectionType,
+                capabilities = Capabilities.Persistent
+            ).toStorageKey()
+        ).isEqualTo(
+            CreateableStorageKey("people", Capabilities.Persistent)
+        )
+    }
+
+    @Test
+    fun handleToStorageKey_noStorageKeyAndNoCapabilities() {
+        assertThat(
+            Recipe.Handle(
+                name = "people",
+                fate = Fate.CREATE,
+                type = personCollectionType
+            ).toStorageKey()
+        ).isEqualTo(
+            CreateableStorageKey("people")
+        )
+    }
+
+    @Test
+    fun particleToPlanParticle() {
+        val handle = Recipe.Handle(
+            name = "contacts",
+            fate = Fate.CREATE,
+            type = contactCollectionType
+        )
+
+        val spec = ParticleSpec(
+            name = "ParticleName",
+            location = "com.Particle",
+            connections = listOf(
+                HandleConnectionSpec("data", HandleMode.Read, contactCollectionType)
+            ).associateBy { it.name }
+        )
+
+        assertThat(
+            Recipe.Particle(
+                spec = spec,
+                handleConnections = listOf(
+                    Recipe.Particle.HandleConnection(
+                        spec = spec.connections["data"]!!,
```

Comment:
How about the following? If ok, change other similar uses too?
```suggestion
                        spec = requireNotNull(spec.connections["data"]),
```

Reply:
Thanks, done :)